dev/devicelab/lib/tasks/run_tests.dart

```diff
@@ -0,0 +1,158 @@
+
+    _findNextMatcherInList(
+      stdout,
+      (String line) => line == 'Application finished.',
+      'Application finished.',
+    );
+
+    return TaskResult.success(null);
+  }
+}
+
+/// Test that the output of `flutter run` is expected.
+abstract class RunOutputTask {
+  RunOutputTask(
+    this.testDirectory,
+    this.testTarget, {
+      required this.release,
+    }
+  );
+
+  /// The directory where the app under test is defined.
+  final String testDirectory;
+  /// The main entry-point file of the application, as run on the device.
+  final String testTarget;
+  /// Whether to run the app in release mode.
+  final bool release;
+
+  Future<TaskResult> call() {
+    return inDirectory<TaskResult>(testDirectory, () async {
+      final Device device = await devices.workingDevice;
+      await device.unlock();
+      final String deviceId = device.deviceId;
+
+      final Completer<void> ready = Completer<void>();
+      final List<String> stdout = <String>[];
+      final List<String> stderr = <String>[];
+
+      final List<String> options = <String>[
+        testTarget,
+        '-d',
+        deviceId,
+        if (release) '--release',
+      ];
+
+      final Process run = await startFlutter(
+        'run',
+        options: options,
+        isBot: false,
+      );
+
+      int? runExitCode;
+      run.stdout
+        .transform<String>(utf8.decoder)
+        .transform<String>(const LineSplitter())
+        .listen((String line) {
+          print('run:stdout: $line');
+          stdout.add(line);
+          if (line.contains('Quit (terminate the application on the device).')) {
+            ready.complete();
+          }
+        });
+      run.stderr
+        .transform<String>(utf8.decoder)
+        .transform<String>(const LineSplitter())
+        .listen((String line) {
+          print('run:stderr: $line');
+          stderr.add(line);
+        });
+      unawaited(run.exitCode.then<void>((int exitCode) { runExitCode = exitCode; }));
+      await Future.any<dynamic>(<Future<dynamic>>[ ready.future, run.exitCode ]);
+      if (runExitCode != null) {
+        throw 'Failed to run test app; runner unexpected exited, with exit code $runExitCode.';
+      }
+      run.stdin.write('q');
+
+      await run.exitCode;
+
+      return verify(stdout, stderr);
+    });
+  }
+
+  /// Verify the output of `flutter run`.
+  TaskResult verify(List<String> stdout, List<String> stderr) => throw UnimplementedError('verify is not implemented');
+
+  /// Helper that verifies a line in [list] matches [matcher].
+  /// The [list] is updated to contain the lines remaining after the match.
+  void _findNextMatcherInList(
+    List<String> list,
+    bool Function(String testLine) matcher,
+    String errorMessageExpectedLine
+  ) {
+    final List<String> copyOfListForErrorMessage = List<String>.from(list);
+
+    while (list.isNotEmpty) {
+      final String nextLine = list.first;
+      list.removeAt(0);
+
+      if (matcher(nextLine)) {
```

Comment:
This may cause out-of-band failures when Xcode is updated, since Xcode loves to add debugging spew, and the macOS build output filtering is very manual:
https://github.com/flutter/flutter/blob/850f3b37d4d2b13d09937836c37dfc12a332b793/packages/flutter_tools/lib/src/macos/build_macos.dart#L22-L28

(as opposed to iOS where the tool output comes from parsing the Xcode build result, not from Xcode build stdout, would really really love macOS to adopt this)
https://github.com/flutter/flutter/blob/8aad1190b7235b93c7f25bc2acbc325820c85384/packages/flutter_tools/lib/src/ios/mac.dart#L406-L409

Either we can make this more flexible and accept any upcoming line, or we can accept that additional filters may need to be applied as part of updating Xcode.

Reply:
I don't think this is a problem as this test ignores unmatched lines. It only checks that the expected lines are in standard output in the correct order.

In fact, there's currently a bunch of [XCode log spew](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8798478644855370033/+/u/run_run_release_test_macos/test_stdout) and the [test still passed](https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20run_release_test_macos/6933/overview).

Please let me know if I misunderstood your concern!

Reply:
> Please let me know if I misunderstood your concern!

Nah I just have Friday brain, your logic LGTM.